File: apps/api/src/main/java/az/ideanest/analytics/application/PlatformAnalytics.java
package az.ideanest.analytics.application;

import az.ideanest.shared.money.Money;
import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

/**
 * The platform's own figures — §4.11's AD-13, issue #313.
 *
 * <p>#313 was blocked on "#95 aggregates one campaign, not the platform", and that is
 * exactly right: {@code ProjectAnalytics} answers a creator asking how their campaign is
 * doing, and every query behind it is filtered by {@code project_id}. What this answers is
 * a different question with the same nouns — how the platform is doing — and the rows it
 * reads are the same daily rollups, summed the other way.
 *
 * <p><strong>Every figure here is derived from rollups that already exist.</strong> That is
 * what made this unblockable without new tables: V27 writes one row per campaign per day,
 * and "volume across the platform" is that table grouped by day instead of by campaign.
 *
 * @param from the first day counted, inclusive
 * @param to the last, inclusive
 * @param computedAt when this was read, so a stale tab is not mistaken for a live figure
 * @param totals the headline numbers over the whole window
 * @param daily one point per day, for the chart
 * @param outcomes how campaigns that closed in the window ended — §4.11's "success rate"
 */
public record PlatformAnalytics(
        LocalDate from,
        LocalDate to,
        Instant computedAt,
        Totals totals,
        List<DailyPoint> daily,
        Outcomes outcomes) {

    /**
     * The headline figures.
     *
     * @param pledgeCount how many pledges were made in the window
     * @param volume what they came to. §21.2 gives nothing to convert with, so this is
     *     denominated in the platform currency and campaigns in another are counted in
     *     {@link #otherCurrencyPledges} rather than converted — a total that silently
     *     added two currencies would be a number nobody could reconcile
     * @param averagePledge {@code volume / pledgeCount}, or zero when nothing was pledged.
     *     Computed here rather than in the browser because dividing money is exactly what
     *     {@link Money} refuses to let callers do carelessly
     * @param backerCount distinct accounts that pledged. Not the same as
     *     {@code pledgeCount}: a backer who pledges to three campaigns is three pledges and
     *     one person, and conflating them overstates reach by whatever the platform's
     *     repeat rate is
     * @param liveProjects campaigns that were {@code LIVE} at the end of the window
     * @param otherCurrencyPledges pledges excluded from {@link #volume} because they were
     *     in another currency. Reported rather than dropped: a figure with a silent
     *     exclusion is worse than one with a stated gap
     */
    public record Totals(
            long pledgeCount,
            Money volume,
            Money averagePledge,
            long backerCount,
            long liveProjects,
            long otherCurrencyPledges) {
    }

    /**
     * One day.
     *
     * @param day the calendar day, in the platform's reporting time zone. V27 stores the
     *     zone on the row for the same reason: a day is not a fixed number of hours from
     *     anybody's point of view except the one that named it
     */
    public record DailyPoint(LocalDate day, long pledgeCount, Money volume) {
    }

    /**
     * How campaigns that closed in the window ended.
     *
     * <p>§4.11 asks for a "success rate", and this is the honest form of it: the two counts
     * and the campaigns still running, rather than one percentage. A single rate would have
     * to decide what to do with campaigns that are neither — and every choice it could make
     * is wrong in a month where a lot of campaigns launched.
     *
     * @param succeeded reached the goal by the deadline
     * @param failed did not
     * @param successRate the fraction, or null when nothing closed in the window. Null
     *     rather than zero, because "no campaigns closed" and "none of them succeeded" are
     *     different facts and a zero on a dashboard reads as the second
     */
    public record Outcomes(long succeeded, long failed, Double successRate) {

        /** The two counts and the rate they imply. */
        public static Outcomes of(long succeeded, long failed) {
            long closed = succeeded + failed;
            return new Outcomes(succeeded, failed, closed == 0 ? null : (double) succeeded / closed);
        }
    }
}

File: apps/web/src/lib/admin/platform-analytics.ts
import { authorizedFetch } from '../api/client';
import { errorFrom } from '../api/problem';
import type { Money } from '../money';

/**
 * §4.11's AD-13: the platform's own figures — issue #313.
 *
 * <h2>Not the campaign dashboard</h2>
 *
 * `/v1/projects/{id}/analytics` is #95's and answers a creator asking about their own
 * campaign. This answers the platform's question with the same nouns, over the same daily
 * rollups summed the other way — which is what made #313 unblockable without a new table.
 *
 * <h2>What it does not answer, and says so</h2>
 *
 * AD-13 lists "volume, success rate, average pledge, cohorts, funnels". The first three are
 * here. Cohorts and funnels are not, and `notBuilt` carries the reason from the service so
 * the screen renders it rather than hard-coding a sentence that would go stale. Approximating
 * them would put two numbers on a dashboard that look like cohorts and funnels and are not.
 */

export interface PlatformTotals {
  pledgeCount: number;
  volume: Money;
  /** Divided by the service, on BigDecimal — dividing money in a browser is how a total drifts. */
  averagePledge: Money;
  /** Distinct accounts. Not the same as `pledgeCount`, and conflating them overstates reach. */
  backerCount: number;
  liveProjects: number;
  /**
   * Pledges left out of `volume` because they were in another currency.
   *
   * Reported rather than dropped: §21.2 gives nothing to convert with, and a total with a
   * silent exclusion is worse than one with a stated gap.
   */
  otherCurrencyPledges: number;
}

export interface PlatformDailyPoint {
  day: string;
  pledgeCount: number;
  volume: Money;
}

export interface PlatformOutcomes {
  succeeded: number;
  failed: number;
  /**
   * The fraction, or null when nothing closed in the window.
   *
   * Null rather than zero: "no campaigns closed" and "none of them succeeded" are different
   * facts, and a zero on a dashboard reads as the second.
   */
  successRate?: number | null;
}

export interface PlatformAnalytics {
  from: string;
  to: string;
  /** When this was read. A stale tab is not a live figure. */
  computedAt: string;
  totals: PlatformTotals;
  daily: PlatformDailyPoint[];
  outcomes: PlatformOutcomes;
  /** What AD-13 asks for and this screen does not answer, in the service's words. */
  notBuilt: string[];
}

/** Volume, success rate and average pledge over a window. */
export async function readPlatformAnalytics(
  from: string | null,
  to: string | null,
  signal?: AbortSignal,
): Promise<PlatformAnalytics> {
  const parameters = new URLSearchParams();
  if (from != null) parameters.set('from', from);
  if (to != null) parameters.set('to', to);

  const query = parameters.toString();
  const response = await authorizedFetch(`/v1/admin/analytics${query === '' ? '' : `?${query}`}`, {
    signal,
  });
  if (!response.ok) throw await errorFrom(response);

  return (await response.json()) as PlatformAnalytics;
}

/**
 * The tallest daily volume in a window, for scaling a chart.
 *
 * Returns zero for an empty window rather than negative infinity, so a caller dividing by it
 * gets a flat chart instead of a page of `NaN`.
 */
export function peakVolume(analytics: PlatformAnalytics): number {
  let peak = 0;
  for (const point of analytics.daily) {
    const value = Number(point.volume.amount);
    if (Number.isFinite(value) && value > peak) peak = value;
  }
  return peak;
}
